base/linalg/dense.jl

```diff
@@ -358,6 +358,10 @@ kron(a::AbstractVector, b::AbstractMatrix) = kron(reshape(a, length(a), 1), b)
 
 # Matrix power
 (^)(A::AbstractMatrix, p::Integer) = p < 0 ? Base.power_by_squaring(inv(A), -p) : Base.power_by_squaring(A, p)
+function (^)(A::AbstractMatrix{T}, p::Integer) where T<:Integer
+    TT = Base.promote_op(^, T, typeof(p))
```

Comment:
I see.

Below you can just write `Base.power_by_squaring(convert(AbstractMatrix{TT}, A), p)`. Then it should be ready to go.